rfcs/007-bytecode-interpreter.md

```diff
@@ -0,0 +1,834 @@
+arguments, return addresses, update frame, local variables, and others.
+
+The basic registers are:
+
+- `next_instruction : CodeAddress`: the address of the next instruction to
+    execute (program counter)
+- `current_closure : Option<Closure>`: a pointer to the closure being evaluated,
+    if any. The VM can offset from there to access the closure's environment.
+    Corresponds to OCaml's `ENV` or Haskell's `Node`.
+
+More registers might be added (such as one or several accumulators or working
+registers) as the implementation progresses.
+
+#### Environment and scopes
+
+The environment is split between a global environment (with the stdlib loaded),
+the stack (function arguments and local let bindings) and the closures'
+environment for captured variables. A scope correspond to either a function or a
+thunk and is introduced by:
+
+- a function definition: `fun <arg> => <body>`
+- argument to a function call: `f (let y = 1 in y)`
+- the bound expression in a let-binding: `let <var> = <bound> in <cont>`
+- array elements: `[<e1>, <e2>, ...]`
+- record fields: `{<field1> = <e1>, <field2> = <e2>, ...}`
+
+Note that all those constructs don't necessarily push stuff on the stack or
+incur the allocation of a thunk, typically when they're trivial (constants,
+variables, etc.). Still, they all formally introduce a new scope.
+
+When entering a scope, the compiler will push the local definitions to the stack
+as they come. Those definitions are then cleaned when leaving the scope, which
+is either at the end of an expression (a `ret`) or when entering a thunk in head
+position, the latter being equivalent to a tail call that can't return (in the
+classical assembly sense of returning, not our value-returning `ret`).
+
+When building a thunk, captured variables are copied from the stack to the
+closure. If some captured variable come from an even higher scope, it is copied
+from the defining scope down to each intermediate scope as a captured variable
+as well: for example, in `fun x => let y = 1 in fun _ => let z = y + 1 in z`,
+the `x` argument will be copied to both the closure `fun y => let z = y + 1 in
+z` and to the closure `y + 1` bound to `z`, to ensure that when we finally get
+to build `y + 1`, `x` is still alive.
+
+#### Instruction set
+
+We don't want to fix the instruction set in this proposal, as it will probably
+evolve as we implement the bytecode compiler and the virtual machine. The
+instruction set should be relatively high-level and standard, handling stack
+operations, function application, variable access, jumps and tests, primitive
+operation call, closure constructions, thunks update, and primitive operations.
+
+### Intermediate representations
+
+We propose the usage of two intermediate representations.
+
+#### AST
+
+The first one is an AST and would more or less correspond to the current unique
+representation, minus runtime-specific constructors. We could have gone closer
```

Comment:
I think we can also get rid of some efficiency-oriented duplication in the current representation, like the distinction between `LetPattern`/`Let` and `RecRecord`/`Record`. Getting rid of these would be convenient for both the LSP and the typechecker, I think.

Reply:
Agreed. In fact I started to draft the first representation and did get rid of the `Let` and `Fun` to keep only the pattern variants.

Reply:
In the end I've mentioned this simplification explicitly.